File: nim/src/math/math.nim
## Module Math that contains the type Math with its procedures to calculate some mathematical problems.
##
## ===========
## Basic Usage
## ===========
##
runnableExamples:
    let mathObj = newMath()
    let fibonacci10 = mathObj.fibonacci(10'u32)
    assert 55'u64 == fibonacci10


type
    Math = ref object


proc newMath*(): Math =
    ## Procedure responsible to create a Math object to call the procedures of Math type (constructor).
    runnableExamples:
        let mathObj = newMath()
    
    return Math()

proc fibonacci*(self: Math, num: uint32): uint64 =
    ## Returns the fibonacci number of `num`.
    runnableExamples:
        let mathObj = newMath()
        let fibonacci10 = mathObj.fibonacci(10'u32)
        assert 55'u64 == fibonacci10

    if num <= 0:
        return 0'u64

    if num == 1:
        return 1'u64

    var fibonacci = @[0'u64,1'u64]
    for i in 2..num:
        try:
            let sum = fibonacci[i - 1] + fibonacci[i - 2]
            fibonacci.add(sum)

        except Exception as ex:
            echo ex.getStackTrace()
            return 0'u64

    return fibonacci[^1]

proc square*(self: Math, num: uint32): uint64 =
    ## Returns the square number of `num`.
    runnableExamples:
        let mathObj = newMath()
        let square10 = mathObj.square(10'u32)
        assert 100'u64 == square10

    return num * num

File: go/src/math/math.go
// Module Math that contains the struct Math with its functions to calculate some mathematical problems.
//
// 	math := math.NewMath()
// 	fmt.Println(math.Fibonacci(10))
package math


// Struct Math that will be used to calculate some mathematical problems.
type Math struct {}


// Function responsible to create a Math object to call the functions of Math struct (constructor).
func NewMath() *Math {
	return new(Math)
}

// Returns the fibonacci number of `num`.
func (math *Math) Fibonacci(num uint32) uint64 {
	if num == 0 {
		return 0
	}

	if num == 1 {
		return 1
	}

	fibonacci := []uint64{0,1}

	for i := 2; i <= int(num); i++ {
		sum := fibonacci[i - 1] + fibonacci[i - 2]
		fibonacci = append(fibonacci, sum)
	}
	return fibonacci[len(fibonacci) - 1]
}

// Returns the square number of `num`.
func (math *Math) Square(num uint32) uint64 {
	return uint64(num * num)
}